src/Adapter/Product/Update/ProductDuplicator.php

```diff
@@ -646,4 +667,123 @@ private function duplicateRelation(array $staticCallback, array $arguments, int
+     *
+     * @param string $table
+     * @param int $oldProductId
+     * @param int $newProductId
+     * @param int $errorCode
+     *
+     * @throws InvalidArgumentException
+     * @throws CannotDuplicateProductException
+     */
+    private function duplicateProductTable(string $table, int $oldProductId, int $newProductId, int $errorCode): void
+    {
+        $oldRows = $this->getRows($table, ['id_product' => $oldProductId], $errorCode);
+        if (empty($oldRows)) {
+            return;
+        }
+        $newRows = $this->replaceInRows($oldRows, ['id_product' => $newProductId]);
+        $this->bulkInsert($table, $newRows, $errorCode);
+    }
+
+    /**
+     * Bulk insert some row values, all row must be formatted with the exact same keys and in the same order
+     * so that the defined column match the values for each row.
+     *
+     * @param string $table
+     * @param array $rowValues
+     * @param int $errorCode
+     */
+    private function bulkInsert(string $table, array $rowValues, int $errorCode): void
+    {
+        if (empty($rowValues)) {
+            return;
+        }
+
+        $insertKeys = array_keys(reset($rowValues));
+        $bulkInsertSql = 'INSERT IGNORE INTO ' . $this->dbPrefix . $table . ' (' . implode(',', $insertKeys) . ') VALUES ';
+        foreach ($rowValues as $i => $rowValue) {
+            if (array_keys($rowValue) !== $insertKeys) {
+                throw new InvalidArgumentException('The provided data has different keys in some rows');
+            }
+
+            $bulkInsertSql .= '(' . implode(',', $rowValue) . ')';
+            if ($i < count($rowValues) - 1) {
+                $bulkInsertSql .= ',';
+            } else {
+                $bulkInsertSql .= ';';
+            }
+        }
+
+        try {
+            $this->connection->executeStatement($bulkInsertSql);
+        } catch (Exception $e) {
+            throw new CannotDuplicateProductException(
+                sprintf('Cannot bulk insert into table %s failed', $table),
+                $errorCode
+            );
+        }
+    }
+
+    /**
+     * Replace columns values in every row.
+     *
+     * @param array $rows
+     * @param array $replacements
+     *
+     * @return array
+     */
+    private function replaceInRows(array $rows, array $replacements): array
+    {
+        $replacedRows = [];
+        foreach ($rows as $key => $row) {
+            $replacedRows[$key] = array_merge($row, $replacements);
+        }
+
+        return $replacedRows;
+    }
+
+    /**
+     * Returns all the columns of a specific table, you can add criteria to filter, prefix is automatically added.
+     *
+     * @param string $table
+     * @param array $criteria
+     * @param int $errorCode
+     * @param array<string, string> $orderBy
+     *
+     * @return array
+     */
+    private function getRows(string $table, array $criteria, int $errorCode, array $orderBy = []): array
```

Comment:
This is way too generic (only the exception is not), this should be part of a trait so it can be reused.

Reply:
The purpose of this was not to do generic code, that's why it's a private method inside the service If we start having to handle this duplicate feature on other entities it will be worth having a dedicated component no problem I'm gonna need to use this method in all the remaining methods that need to be refactored in the second PR that's why I extracted this in a dedicated private method instead of duplicating and having a code less readable

But for now no need for a trait that will not be used anywhere else. If you have a suggestion for a better naming though I'm ok! ut since the method purpose and usage IS generic I only had generic name in mind, I can propose `getTableRows` but that's the best I have 😅 And I'm not sure it will please you much better 😛